### DISCUSSION TITLE
Pico W Wifi Troubles

### DISCUSSION THREAD

**Discussion by edmcman** [micropython/micropython#16471] [RP2040 / Pico]
I'm having trouble with the Pico W connecting to Wifi.  It appears that if it is unused for a while, the Pico disconnects but can't seem to detect this.

I get weird debug messages when this happens:

```
[ 5322065] ASYNC(0000,49,0,0,0)
```

If I make a request every 5 minutes, the problem does not happen.

When the Pico disconnects, `network.isconnected()` returns True and `network.status()` returns 3.

Any idea what is going on?  Is my pico just flaky perhaps?

**Comment by edmcman** [comment]
Current code located here https://github.com/edmcman/rpi-temp-server/tree/debug

**Reply by mendenm** [reply]
My solution to this was to install uping, and just ping the network every so often.  It's a hack, but made connections stable for months.

**Reply by davefes** [reply]
I was interested in your solution so checked your GitHub project.  I don't see anything in `main.py` regarding uping.  How do you get it to work?

Curious about the sequence of events:

> but made connections stable for months.

Have you used this before on other projects?

**Reply by mendenm** [reply]
here is the main loop I used for a powerline phase monitor, along with ancillary code.  The ```keep_net_alive()``` function is what you may be interested in.  It's not pretty, but it works.


```
#test microdot

from microdot_asyncio import Microdot
import asyncio
import uping

app = Microdot()

@app.get('/')
async def index(request):
    #print(request.client_addr, request.body)
    return '''
<html>
<head> </head>
<body>
<h1>Hello, World!</h1>
</body>
</html>
''', 200, {'Content-Type': 'text/html'}


@app.get('/set_time')
async def set_time(request):
    #print(request.client_addr, request.body)
    import ntptime
    ntptime.timeout = 10
    ntptime.settime()
    from machine import RTC
    rtc = RTC()
    return f'''
<html>
<head> </head>
<body>
<h1>{str(rtc.datetime()):s} UTC</h1>
</body>
</html>
''', 200, {'Content-Type': 'text/html'}


@app.get('/index/')
def index_files(request):
    import os
    top = '''
<html>
<head> </head>
<body>
<ol>
'''
    vals = '\n'.join( "<li>"+x+"</li>" for x in sorted(os.listdir("/timelogs/")) )
    
    bot = '''
</ol>
</body>
</html>
'''
    return top + vals + bot, 200, {"Content-type":"text/html"}

@app.get('/timelogs/<string:filename>')
def handle_file(request, filename = None):
    from microdot_asyncio import send_file
    args = request.args
    delete = 'delete' in args and args['delete'][0] == '1'
            
    try:
        if delete:
            import os
            os.remove('/timelogs/' + filename)
            return 'removed', 200, { 'Content-Type':'text/plain' }
        else:
            return send_file('/timelogs/' + filename, content_type = 'text/plain',
                 extra_headers = {'Content-Disposition':  'attachment'}
                         )
    except OSError:
        return "File not found:" + filename, 404, {}
    
from detect_power_line import doit_events, current_status

@app.get('/status')
def index_files(request):
    import os
    from machine import RTC
    rtc = RTC()
    
    top = '''
<html>
<meta charset="UTF-8">
<style>
table {
  border-collapse: collapse;
}
th, td {
  padding-right: 5px;
  padding-left: 5px;
  padding-top: 5px;
}
th {
  text-align: center;
  border: 2px solid black;
}
td {
  text-align: right;
  border: 1px solid black;
}
</style>
'''
    head = '''
<head> Date: {0:04d}-{1:02d}-{2:02d} {4:02d}:{5:02d}:{6:02d}</head>
'''.format(*rtc.datetime())

    body = '''
<body>
<table>

<tr>
<th> time / ms </th>
<th> ADC </th>
<th> IQ mag / mLSB </th>
<th> I </th> <th> Q </th> <th> φ/° </th>
<th> RMS / LSB </th>
<th> Δt / s </th>
</tr>
'''
    
    bot = '''
<table>
</body>
</html>
'''

    vals = '\n'.join(
        [ '\n'.join(['<tr>'] + ["<td>" + x+ "</td>" for x in r.split()] + ['</tr>'])  for r in current_status ]
    )
    
    return top + head + body + vals + bot, 200, {"Content-type":"text/html", "Refresh":"10"}

nic = None

netdict = json.load(open("networks.json","r"))

async def keep_net_alive(verbose):
    global nic
    import network
    last_status = -1
    while nic != "STOP":
        stat = nic.status() if nic is not None else -4
        if stat < 0:
            if verbose and nic is not None: print("no net:", nic.status(), nic.ifconfig())
            if nic is not None: nic.active(False)
            last_status = stat
            #network hard crashed...
            network.hostname("powerline_pico")
            network.country('US')
            #ap = network.WLAN(network.AP_IF)
            #ap.active(False) #turn off any AP interface activity
            nic = network.WLAN(network.STA_IF)
            nic.active(True)
            await asyncio.sleep(1)
            if verbose: print("attempted activation:", nic.status())
        elif last_status < 0 or not nic.isconnected(): #alive, but not connected
            nets = nic.scan()
            if verbose: print("found", [str(x[0],'utf8') for x in nets])
            for net in nets:
                netname = str(net[0],"utf8") #might be bytes from NIC
                if netname in netdict:
                    nic.disconnect() #start  over here
                    nic.active(False)
                    await asyncio.sleep(1)
                    nic.active(True)
                    nic.connect(net[0], netdict[netname])
                    while nic.status() > network.STAT_IDLE and nic.status() != network.STAT_GOT_IP: 
                        await asyncio.sleep(1) #check again later
                    if verbose: print("connected to", str(net[0], 'utf8'), nic.ifconfig(), nic.config('hostname'))
                    break

            nic.config(hostname  = "powerline_pico")
            last_status = nic.status()
            await asyncio.sleep(1) #check again later
        else:
            config =  nic.ifconfig()
            uping.ping(config[-1], count = 1)
            if verbose: print("on net:", nic.status(), nic.ifconfig())
            await asyncio.sleep(5)            

async def set_initial_time(log):
    import ntptime
    while 1:
        try:
            if log: print("setting time")
            ntptime.timeout = 5
            ntptime.settime()
            if log:
                from machine import RTC
                rtc = RTC()
                print("set time to:", rtc.datetime())
            break
        except OSError as e:
            print("oops", str(e))
            await asyncio.sleep(5)
            pass #just keep looping until it happens
    
async def doit(log = False, startidx = 0):
    global nic
    #print("starting line monitor")
    net_monitor = asyncio.create_task(keep_net_alive(log))
    
    time_init = asyncio.create_task(set_initial_time(log)) #should return after time set, when network is up
    
    line_monitor = asyncio.create_task(doit_events(log = log))
    #print("starting web  server")
    web_server = asyncio.create_task(app.start_server(host='0.0.0.0', port=5000, debug = False, ssl = None))
    await line_monitor
    app.shutdown()
    nic = "STOP"
    await net_monitor
    
if __name__=="__main__":
    asyncio.run(doit(log=True))
```

**Reply by davefes** [reply]
Thanks.  I will have a look at that as it could eliminate a `ping client` and watchdogs on my existing server.

**Comment by edmcman** [comment]
I'm running this fish script: `set sleep 1; while true; date; echo "Sleeping $sleep"; sleep $s
leep; wget -q http://192.168.1.4/temp -O /dev/stdout || break; set sleep (math "2*$sleep"); end ; not
ify done`

It's currently past 4096 and sleeping 8192.

Hard to say because it is non-deterministic, but I did just add `wlan.config(pm=network.WLAN.PM_NONE)`.  Maybe that's the solution. Which still feels like a bug.

**Reply by davefes** [reply]
Note: I only have ESP32 WiFi experience.

Some other hints found floating around:

> config(timeout_ms=xx), PM_NONE and txpower (set to some value)

The only way I have been able to maintain a reliable connection for an alarm server to a hotspot dongle  is to run a ping client  and if the server doesn't process the ping within a certain time it re-boots.  Likewise, if the ping client doesn't receive an ack, then it will re-boot as well.

You then have to handle re-connects properly.

**Reply by edmcman** [reply]
So with PM_PERFORMANCE it failed after 1024 seconds. Trying again with PM_NONE.

It's sad to hear how many people have had to resort to ping hacks!

**Reply by edmcman** [reply]
On PM_NONE, I'm at 4096 seconds again.  I'll try to let it run some more.

Out of curiosity, has anyone else that has faced this problem tried PM_NONE?

**Reply by davefes** [reply]
The two guys that have helped in the past are:
`glenn20` and `shariltumin`
I recall they both mentioned `PM_NONE'

That suggestion happened after I got the `ping client` setup working and couldn't be bothered, at the time, to try it.

I usually takes months for connect failure so I got tired of testing.

`hippy` on the Pico Micropython forum might be worthwhile searching or having a chat with.

Good luck!

**Reply by davefes** [reply]
Here is my latest `wifi_functions.py` script, tested but not been used at the moment:
```py
import network
from time import sleep
import config


network.country("NZ")   # New Zealand
sta_if = network.WLAN(network.STA_IF)

sta_if.active(False)  # TOGGLING THIS FIRST PREVENTS "OSError: Wifi Internal Error"
sta_if.active(True)

#sta_if.config(pm=sta_if.PM_NONE)  # disable power management, seems to need active(TRUE) first
sta_if.config(pm=sta_if.PM_PERFORMANCE)   # this works fine for me, and saves power/heat

sta_if.config(txpower=10)  # Fixes wifi ESP32 C3 5x5 WS2812 boards
sta_if.connect(config.hotspot, config.password")

print('Connection pending')

while not sta_if.isconnected():
    if sta_if.status() == 1000:
        print('Waiting... status: {}...'.format(sta_if.status()))
    else: 
        print('Status: {}...'.format(sta_if.status()))

    sleep(1)
   
print('Completed with {}...'.format(sta_if.status()))  # 1010 is connected.

if sta_if.status() == 1010:
    print('Wifi connection successful')
```

**Comment by edmcman** [comment]
Thanks all.  I also just found https://github.com/micropython/micropython/issues/9455 which seems like the same problem.  I will post more about the issue there.

**Reply by scruss** [reply]
There's been a lot of work done on network_cyw43 since that issue was posted, so it might not be best to revisit that one

**Reply by edmcman** [reply]
Thanks, I'll make a new issue when I am ready then.  I'm running tests and collecting data right now.

**Comment by edmcman** [comment]
I'm still continuing to test.  I am less sure that the PM fixes the issue.  I was testing using a laptop, and I think a significant factor may have been moving or flipping the pico.  I am moving to an isolated computer so this won't be an issue anymore!

**Reply by scruss** [reply]
Maybe expecting a wifi micro-controller to be always available on the network is a bit of a stretch. The MQTT model of the microcontroller (client) waking up periodically, taking a reading, connecting to the network and sending that reading to a broker is pretty robust. Peter Hinch's [MicroPython Asynchronous MQTT](https://github.com/peterhinch/micropython-mqtt/tree/master) is remarkably resilient.

But MQTT is of no use if you need to query your sensor *right now*

**Reply by edmcman** [reply]
> Maybe expecting a wifi micro-controller to be always available on the network is a bit of a stretch. 

What do you mean by this?  I'm not expecting 100% availability, but I do expect that if the micro-controller gets knocked offline or drops a few packets, it is able to recover.  Just like most devices that are on wifi.

What is happening now is some unknown cause -- my current guess is maybe an AP disconnect or dropped packets -- causes the micro-controller to enter a state where it disconnects from the wifi but reports that it is still connected to micropython.  It never recovers.  I consider this a bug (and unacceptable, even for a micro-controller).  Hopefully that's not under dispute!

**Reply by scruss** [reply]
I'm not disputing that falling off a network yet still reporting connected is unacceptable. It's really annoying.

What I was suggesting was that — in my experience — networked micro-controllers haven't typically been used in an always-on/always-available configuration.  They've typically taken a measurement, connected to a network, sent the measurement, then disconnected. This is a fairly robust model, as the connection is opened then data is sent almost immediately.

I don't know what a Pico W does in the background to keep its connection open. Clearly, not enough to stay connected to some APs

**Reply by edmcman** [reply]
My son also suggested a similar approach when I told him about the problem.  This is a personal project for me, and I would like to find out and hopefully fix whatever is going wrong so that others don't encounter the same problem.  If that proves untenable, I might switch to storing measurements in a more stable location.

**Comment by edmcman** [comment]
I've been running experiments for almost a day now, with the power management choice selected randomly from: `["None", "0xa11140", "network.WLAN.PM_NONE", "network.WLAN.PM_PERFORMANCE", "network.WLAN.PM_POWERSAVE"]`

I limited each test to intervals of no more than 2048 seconds.  I will probably try longer intervals later.

We are currently on the 18th test.  Out of the 17 that completed, 8 failed.  Here is the breakdown of PMs for the failures:

      6 network.WLAN.PM_POWERSAVE
      2 None

But overall the PMs aren't quite well distributed yet:

      3 0xa11140
      6 network.WLAN.PM_NONE
      7 network.WLAN.PM_POWERSAVE
      3 None

But I think for me the big take away is that the problem _does_ occur with `None` (the default PM mode).  It is too early to say for sure, but 0xa11140 and `PM_NONE` might fix or make the problem less likely.

**Reply by edmcman** [reply]
The `None` bit is interesting because I believe micropython adopted the default specifically because it was more reliable, though I don't have the issue link handy now.

**Comment by edmcman** [comment]
Since I have enough data to make a coherent issue, I did so here: https://github.com/micropython/micropython/issues/16482

I will close this discussion in favor of this issue.  Thank you all for the input and help so far.
